x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -0,0 +1,418 @@
+import { RawRule } from '../types';
+import { RuleRunMetricsStore } from '../lib/rule_run_metrics_store';
+import { injectActionParams } from './inject_action_params';
+import { ExecutionHandlerOptions, RuleTaskInstance } from './types';
+import { TaskRunnerContext } from './task_runner_factory';
+import { transformActionParams } from './transform_action_params';
+import { Alert } from '../alert';
+import { NormalizedRuleType } from '../rule_type_registry';
+import {
+  ActionsCompletion,
+  AlertInstanceContext,
+  AlertInstanceState,
+  RuleAction,
+  RuleTypeParams,
+  RuleTypeState,
+  SanitizedRule,
+} from '../../common';
+
+export class ExecutionHandler<
+  Params extends RuleTypeParams,
+  ExtractedParams extends RuleTypeParams,
+  RuleState extends RuleTypeState,
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+> {
+  public request: KibanaRequest;
+  private logger: Logger;
+  private alertingEventLogger: PublicMethodsOf<AlertingEventLogger>;
+  private rule: SanitizedRule<Params>;
+  private ruleType: NormalizedRuleType<
+    Params,
+    ExtractedParams,
+    RuleState,
+    State,
+    Context,
+    ActionGroupIds,
+    RecoveryActionGroupId
+  >;
+  private ruleTypeActionGroups;
+  private taskRunnerContext: TaskRunnerContext;
+  private taskInstance: RuleTaskInstance;
+  private ruleRunMetricsStore: RuleRunMetricsStore;
+  private apiKey: RawRule['apiKey'];
+  private ruleConsumer: string;
+  private executionId: string;
+  private ruleLabel: string;
+  private ephemeralActionsToSchedule: number;
+  private mutedAlertIdsSet: Set<string>;
+  private CHUNK_SIZE = 1000;
+  private skippedAlerts: { [key: string]: { reason: string } } = {};
+
+  constructor({
+    rule,
+    ruleType,
+    logger,
+    alertingEventLogger,
+    taskRunnerContext,
+    taskInstance,
+    ruleRunMetricsStore,
+    apiKey,
+    ruleConsumer,
+    executionId,
+    ruleLabel,
+    request,
+  }: ExecutionHandlerOptions<
+    Params,
+    ExtractedParams,
+    RuleState,
+    State,
+    Context,
+    ActionGroupIds,
+    RecoveryActionGroupId
+  >) {
+    this.logger = logger;
+    this.alertingEventLogger = alertingEventLogger;
+    this.rule = rule;
+    this.ruleType = ruleType;
+    this.ruleTypeActionGroups = new Map(
+      ruleType.actionGroups.map((actionGroup) => [actionGroup.id, actionGroup.name])
+    );
+    this.taskRunnerContext = taskRunnerContext;
+    this.taskInstance = taskInstance;
+    this.ruleRunMetricsStore = ruleRunMetricsStore;
+    this.apiKey = apiKey;
+    this.ruleConsumer = ruleConsumer;
+    this.executionId = executionId;
+    this.ruleLabel = ruleLabel;
+    this.request = request;
+    this.ephemeralActionsToSchedule = taskRunnerContext.maxEphemeralActionsPerRule;
+    this.mutedAlertIdsSet = new Set(rule.mutedInstanceIds);
+  }
+
+  public async run(
```

Comment:
WDYT of making this take `activeAlerts` and `recoveredAlerts` instead of having to call it twice with different flags? That way, if we needed to add some other alert grouping (like `noDataAlerts` maybe) in the future, we wouldn't have to be adding and passing more flags.

Reply:
Actually i tried this first but created too much code repeat, but will try again once we are sure that this PR does not break anything :)